src/stores/utils/rewardable.js

```diff
@@ -0,0 +1,92 @@
+import BN from 'bignumber.js';
+import { FEE_MANAGER_MODE } from './bridgeMode'
+
+export const validFee = (fee) => {
+  const zeroBN = new BN(0)
+  return !zeroBN.eq(fee)
+}
+
+export const getFeeAtBlock = (feeArray, blockNumber) => {
+  for (let i = feeArray.length - 1; i >= 0; i--) {
+    if (blockNumber >= feeArray[i].blockNumber) {
+      return feeArray[i].fee
+    }
+  }
+  return new BN(0)
+}
+
+export const getFeeToApply = (homeFeeManager, foreignFeeManager, homeToForeignDirection) => {
+  const data = getRewardableData(homeFeeManager, foreignFeeManager)
+  return homeToForeignDirection ? data.homeFee : data.foreignFee
+}
+
+export const calculateValueFee = (feeList, feeCollected, feeApplied) => {
+  return (event) => {
+    const fee = getFeeAtBlock(feeList, event.blockNumber)
+    let calculatedFee = event.value.multipliedBy(fee)
+    if (feeApplied) {
```

Comment:
Could you elaborate this a bit? What purpose of `feeApplied` is?

Reply:
On `UserRequestForSignature` and `AffirmationCompleted` events, the value could be the initial value that the user made the transfer or could be the value of transfer with the fee already subtracted. For `UserRequestForSignature` only in NATIVE-TO-ERC mode and when Fee Manager is `ONE_DIRECTION` the value is the same as the initial value from transfer, while on `AffirmationCompleted` the value is always the same as the value from user transfer. So the purpose of `feeApplied` is to correctly calculate the fee from the event depending if fee was already applied to the value.

Reply:
Oh! I see. Seems that we duplicate the logic from the contracts, doesn't it?
The final call will be on you but I should (sorry!!!) ask again: will it worth for us to get validator reward information from the special events since it does both: simplify the code on the UI side and allow separate the same logic from the UI to the contracts as so if we change the reward logic in the contracts we don't need to modify the code in the UI?

Reply:
Thanks for asking the questions! I think you are right, adding special events to inform about the taken fees from the transfer will help simplify logic on UI and also avoid duplicating the logic from the contracts. Here are some possible events we can emit:
- `FeeDistributedFromSignatures(uint256 feeAmount, bytes32 indexed txHash)` every time `distributeFeeFromSignatures` is called (on home or foreign)
- `FeeDistributedFromAffirmation(uint256 feeAmount, bytes32 indexed txHash)` every time `distributeFeeFromAffirmation` is called (on home)

Logic on UI to calculate the fee will consist of listening to these events and accumulate the `feeAmount` parameter. I thought of txHash parameter as `indexed` so if for some reason you want to know the fee taken for a transfer, you can listen to the events and filter it by the txHash of the transfer. What do you think about these events names and when to emit them? Should we include any other parameter related to the transfer (recipient, original value transfer, etc)?

Reply:
Should we create an issue for this?

Reply:
Yes, please. As soon as the issue in the contract addressed, these PR should be updated to reflect the changes in the contracts.

Reply:
Created PR https://github.com/poanetwork/poa-bridge-contracts/pull/171. After it is merged, logic to calculate collected fees should be updated

Reply:
Updated the logic to calculate collected fees using the new fee events on https://github.com/poanetwork/bridge-ui/pull/211/commits/9aee28318a741d1729e10e612a920b89f492fe2d.
To be able to also display collected fees for `ETC-WETC` bridge that doesn't have the new fee events and also to correctly display all the fee collected if in the future we upgrade the contracts to a version that contains the fee events, I kept the previous logic to calculate collected fees and it will be executed only if:
- The amount of `UserRequestForSignature` events are greater than the amount of `FeeDistributedFromSignatures` events for Home to Foreign transfers
https://github.com/poanetwork/bridge-ui/blob/9aee28318a741d1729e10e612a920b89f492fe2d/src/stores/HomeStore.js#L456 
- The amount of `AffirmationCompleted` events are greater than the amount of `FeeDistributedFromAffirmation` events for Foreign to Home transfers
https://github.com/poanetwork/bridge-ui/blob/9aee28318a741d1729e10e612a920b89f492fe2d/src/stores/HomeStore.js#L476

Reply:
Should we support both ways to totaling the fees? I understand that existing `ETC-WETC` bridge does not emit events but what if we generate all needed events during the next bridge contract upgrade? For example, we can prepare a method `upgrade` in the contracts for both sides that will be called as part of the upgrade procedure. We can take into account all transfers that already made and reflect them in the `upgrade` method as so it will emit the corresponding number of `FeeDistributedFromAffirmation` on the Home side and similar is for the Foreign side but the event will be `FeeDistributedFromSignatures`. If we do everything properly, we can get overall amount of fees just by getting information from `FeeDistributedFromAffirmation` and `FeeDistributedFromSignatures` events.

Reply:
Right, thanks for the example. In that case, we only need to support calculate collected fees using the new events. I'll remove previous logic since it is not needed

Reply:
Done https://github.com/poanetwork/bridge-ui/pull/211/commits/271d13f341ce3b2afa0d0d1c735c90666b07e98d